app/src/main/java/org/mozilla/focus/search/ManualAddSearchEnginePreference.java

```diff
@@ -0,0 +1,88 @@
+import android.widget.EditText;
+
+import org.mozilla.focus.R;
+import org.mozilla.focus.utils.UrlUtils;
+
+public class ManualAddSearchEnginePreference extends Preference {
+    EditText engineNameEditText;
+    EditText searchQueryEditText;
+    TextInputLayout engineNameErrorLayout;
+    TextInputLayout searchQueryErrorLayout;
+
+    public ManualAddSearchEnginePreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public ManualAddSearchEnginePreference(Context context, AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+    }
+
+    @Override
+    protected View onCreateView(ViewGroup parent) {
+        final View view = super.onCreateView(parent);
+
+        engineNameErrorLayout = view.findViewById(R.id.edit_engine_name_layout);
+        searchQueryErrorLayout = view.findViewById(R.id.edit_search_string_layout);
+
+        engineNameEditText = view.findViewById(R.id.edit_engine_name);
+        engineNameEditText.addTextChangedListener(buildTextWatcherForErrorLayout(engineNameErrorLayout));
+        searchQueryEditText = view.findViewById(R.id.edit_search_string);
+        searchQueryEditText.addTextChangedListener(buildTextWatcherForErrorLayout(searchQueryErrorLayout));
+        return view;
+    }
+
+    private boolean engineNameIsUnique(String engineName) {
+        final SharedPreferences sharedPreferences = getContext().getSharedPreferences(SearchEngineManager.PREF_FILE_SEARCH_ENGINES, Context.MODE_PRIVATE);
+        return !sharedPreferences.contains(engineName);
+    }
+
+    public boolean validateEngineNameAndShowError(String engineName) {
+        if (TextUtils.isEmpty(engineName) || !engineNameIsUnique(engineName)) {
+            engineNameErrorLayout.setError(getContext().getString(R.string.search_add_error_empty_name));
+            return false;
+        } else {
+            engineNameErrorLayout.setError(null);
+            return true;
+        }
+    }
+
+    public boolean validateSearchQueryAndShowError(String searchQuery) {
+        if (TextUtils.isEmpty(searchQuery)) {
+            searchQueryErrorLayout.setError(getContext().getString(R.string.search_add_error_empty_search));
+            return false;
+        } else if (!UrlUtils.isValidSearchQueryUrl(searchQuery)) {
+            searchQueryErrorLayout.setError(getContext().getString(R.string.search_add_error_format));
+            return false;
+        } else {
+            searchQueryErrorLayout.setError(null);
+            return true;
+        }
+    }
+
+    private TextWatcher buildTextWatcherForErrorLayout(final TextInputLayout errorLayout) {
+        return new TextWatcher() {
+            @Override
+            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {}
+
+            @Override
+            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
+                errorLayout.setError(null);
```

Comment:
Oh, is this needed? I've been calling setError() on an EditText somewhere and it seemed to clear automatically when the text was changed. Is this because the error is set on the TextInputLayout?

Reply:
The error state looks a little different when called with setError on the EditText - it's a "!" and a hovertext, and doesn't match the material design docs. I do think the APIs for EditText.setError are nicer, because the error autohides, so maybe there's an update to those APIs that removes the need to add a TextWatcher.